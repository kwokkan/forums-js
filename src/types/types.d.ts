
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
