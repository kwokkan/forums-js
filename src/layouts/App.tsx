import React from "react";
import { Container, Content, Header } from "rsuite";
import { forums } from "../mocks/mockForums";
import { Announcement } from "./Announcement";
import { Forum } from "./Forum";
import { Forums } from "./Forums";
import { Navigation } from "./Navigation";
import { Thread } from "./Thread";

interface IProps {
    title: string;
}

export function App(props: IProps) {
    return (
        <div className="navbar-page">
            <Container>
                <Header>
                    <Navigation title={props.title}></Navigation>
                </Header>

                <Content>
                    <Announcement title="Coming soon" message="Coming soon."></Announcement>

                    <Forums forums={forums}></Forums>

                    {forums.map(x =>
                        <Forum key={x.id} forum={x}></Forum>
                    )}

                    {forums.filter(x => x.threads).flatMap(x => x.threads!).map(x =>
                        <Thread key={x.id} thread={x}></Thread>
                    )}
                </Content>
            </Container>
        </div>
    );
}
