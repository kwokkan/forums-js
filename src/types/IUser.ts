export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
