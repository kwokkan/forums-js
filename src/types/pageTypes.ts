import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
