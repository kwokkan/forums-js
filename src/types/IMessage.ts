import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
