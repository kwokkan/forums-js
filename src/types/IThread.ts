import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
