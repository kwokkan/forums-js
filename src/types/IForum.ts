import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
