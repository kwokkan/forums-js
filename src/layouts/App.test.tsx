/**
 * @jest-environment jsdom
*/
import React from "react";
import { create } from "react-test-renderer";
import { IUser } from "../types/IUser";
import { App } from "./App";

test("Renders with custom child", () => {
    const child = <p>Test child</p>;

    const tree = create(
        <App title="Test app">
            {child}
        </App>
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with custom child and user", () => {
    const user: IUser = {
        id: 1,
        name: "Test user",
        joinedDate: 1
    };

    const child = <p>Test child</p>;

    const tree = create(
        <App title="Test app" user={user}>
            {child}
        </App>
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
