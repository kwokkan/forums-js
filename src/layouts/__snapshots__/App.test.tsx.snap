// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with custom child 1`] = `
<div
  className="navbar-page"
>
  <div
    className="rs-container"
  >
    <div
      className="rs-header"
    >
      <div
        className="rs-navbar rs-navbar-default"
        role="navigation"
      >
        <div
          className="rs-navbar-header"
        >
          <a
            className="navbar-brand logo"
            href="/"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <img
              alt="Logo"
              className="mr-3"
              height="24"
              src="/logo.png"
              title="Test app"
              width="24"
            />
            Test app
          </a>
        </div>
        <div
          className="rs-navbar-body"
        >
          <div
            className="rs-nav rs-nav-default rs-navbar-nav rs-nav-horizontal"
          >
            <ul>
              <li
                className="rs-nav-item"
              >
                <a
                  className="rs-nav-item-content"
                  href="/"
                  onClick={[Function]}
                  onMouseEnter={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <i
                    className="rs-icon rs-icon-home"
                  />
                  Home
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
              <li
                className="rs-nav-item"
              >
                <a
                  className="rs-nav-item-content"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Forums
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
              <li
                className="rs-nav-item"
              >
                <a
                  className="rs-nav-item-content"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  New Posts
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
              <li
                className="rs-dropdown rs-dropdown-placement-bottom-start"
                role="menu"
              >
                <ul
                  className="rs-dropdown-menu"
                  role="menu"
                >
                  <li
                    className="rs-dropdown-item"
                  >
                    <a
                      className="rs-dropdown-item-content"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      Company
                    </a>
                  </li>
                  <li
                    className="rs-dropdown-item"
                  >
                    <a
                      className="rs-dropdown-item-content"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      Team
                    </a>
                  </li>
                  <li
                    className="rs-dropdown-item"
                  >
                    <a
                      className="rs-dropdown-item-content"
                      onClick={[Function]}
                      tabIndex={-1}
                    >
                      Contact
                    </a>
                  </li>
                </ul>
                <a
                  className="rs-btn rs-btn-subtle rs-dropdown-toggle"
                  onClick={[Function]}
                  tabIndex={0}
                >
                  About
                  <span
                    className="rs-dropdown-toggle-caret"
                  />
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
            </ul>
          </div>
          <div
            className="rs-nav rs-nav-default rs-navbar-nav rs-navbar-right rs-nav-horizontal"
          >
            <ul>
              <li
                className="rs-nav-item"
              >
                <a
                  className="rs-nav-item-content"
                  href="https://github.com/kwokkan/forums-js"
                  onClick={[Function]}
                  rel="noreferrer noopener"
                  role="button"
                  tabIndex={0}
                  target="_blank"
                >
                  <i
                    className="rs-icon rs-icon-github"
                  />
                  Fork me
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
              <li
                className="rs-nav-item"
              >
                <a
                  className="rs-nav-item-content"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <i
                    className="rs-icon rs-icon-cog"
                  />
                  Settings
                  <span
                    className="rs-ripple-pond"
                  >
                    <span
                      className="rs-ripple"
                      style={Object {}}
                    />
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      className="rs-content"
    >
      <div
        className="rs-message rs-message-info rs-message-show rs-message-has-title rs-message-has-icon"
      >
        <div
          className="rs-message-container"
        >
          <div
            className="rs-message-icon-wrapper"
          >
            <i
              className="rs-icon rs-icon-info"
            />
          </div>
          <div
            className="rs-message-content"
          >
            <h5
              className="rs-message-header"
            >
              Coming soon
            </h5>
            <div
              className="rs-message-body"
            >
              <p>
                Coming soon.
              </p>
            </div>
          </div>
        </div>
      </div>
      <p>
        Test child
      </p>
    </div>
  </div>
</div>
`;
